### PR TITLE
DEAP Runner

### PR DESCRIPTION
**Pull request by samgdotson**
This PR enables users to run a model with DEAP rather than Pymoo. It also adds some logging capabilities and local file saving.

### CODE REVIEW

File: osier/equations.py [hunk 1]
```diff
@@ -116,24 +116,25 @@ def objective_from_energy(technology_list, attribute, solved_dispatch_model):
...
+    valid_techs = [t for t in technology_list if hasattr(t, attribute)]
+    column_names = get_tech_names(valid_techs)
     dispatch_results = solved_dispatch_model.results
     power_units = solved_dispatch_model.power_units
     time_delta = solved_dispatch_model.time_delta
     dispatch_values = dispatch_results[column_names].values.T*power_units*time_delta
-
-
+    
+    mass_u = valid_techs[0].unit_mass
```
Comment (LukeSeifert) [inline osier/equations.py:126]: How come the first valid tech's unit mass is used? Do they all have the same unit mass at this point?

File: osier/models/capacity_expansion.py [hunk 2]
```diff
@@ -212,14 +210,6 @@ def _evaluate(self, x, out, *args, **kwargs):
...
-            out["G"] = out_constr
-
 
-        if (self.verbosity =='debug'):
-            self.print_problem_formulation()
-
-            print(f"Model solved? {model.results != None}\n")
-
-            print("Objective Values:\n")
-            print(out["F"])
```
Comment (LukeSeifert) [inline osier/models/capacity_expansion.py:225]: How come you didn't keep a debug statement here?

File: osier/models/deap_runner.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+from deap import base
+from deap.benchmarks.tools import igd
+from deap import creator
+from deap import tools
+
+from osier import *
+from osier.tech_library import *
+
+
+def uniform(low, up, size=None):
```
Comment (LukeSeifert) [inline osier/models/deap_runner.py:20]: Missing docstring

File: osier/models/deap_runner.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        The problem to execute. In theory, this could be any object
+        that has a method :meth:`evaluate` (for example, if it inherits 
+        from :class:`pymoo.ElementwiseProblem`) and outputs a dictionary.
+    algorithm : Optional, str
+        Indicates the algorithm to use. Accepts ['nsga2', 'nsga3']. 
+        NSGA2 is recommended for problems with three or fewer objectives.
+        NSGA3 is recommended for problems with three or more objectives. 
+    lower_bound : Optional, float, int
+        The lower bound for each 'gene' in a candidate individual.
+    upper_bound : Optional, float, int
```
Comment (LukeSeifert) [inline osier/models/deap_runner.py:51]: How come `float` and `int` are both given here? Wouldn't just `float` be sufficient?

File: osier/models/deap_runner.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            self.algorithm = 'nsga2'
+        elif self.n_obj >= 3:
+            self.algorithm = 'nsga3' 
+        self.n_dim = self.problem.n_var
+        self.pop_size = pop_size
+        self.repair = repair
+        self.completed_generations = 0
+        self.solve_time = 0.0
+        self.last_population = None
+        self.save_directory = Path(save_directory) if save_directory else None
```
Comment (LukeSeifert) [inline osier/models/deap_runner.py:84]: `save_directory` default appears to be `f"{timestr}-OsierModel.pkl"`, how come you don't define it here instead?

File: osier/models/deap_runner.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+            self.last_population = pop
+            self.completed_generations += 1
+            print(self.logbook.stream)
+
+        end = time.perf_counter()
+        self.solve_time += (end-start)
+
+        return self.last_population, self.logbook, self.pareto_front
+    
+    def save_model(self, fpath=None):
```
Comment (LukeSeifert) [inline osier/models/deap_runner.py:227]: `fpath` appears to be unnecessary here

**Review by LukeSeifert — COMMENTED** [review]
Looks good overall, I just had a few minor comments and questions.

File: osier/equations.py (hunk 1 above)
Comment (samgdotson) [reply osier/equations.py:126]: yes, the units are regularized

File: osier/models/capacity_expansion.py (hunk 2 above)
Comment (samgdotson) [reply osier/models/capacity_expansion.py:225]: I'm going to revamp the logging stuff in a later PR. I kind of half-assed it here, so I decided just to remove it.

File: osier/models/deap_runner.py (hunk 4 above)
Comment (samgdotson) [reply osier/models/deap_runner.py:51]: In most cases, yes. But sometimes you want variables to only take on integer amounts.

File: osier/models/deap_runner.py (hunk 6 above)
Comment (samgdotson) [reply osier/models/deap_runner.py:227]: In the future I want users to be able to pass their own file path that overrides `save_directory`.

File: osier/models/deap_runner.py (hunk 5 above)
Comment (samgdotson) [reply osier/models/deap_runner.py:84]: The default is `None`. The file NAME is `f"{timestr}-OsierModel.pkl"`. It either saves the file in the current directory or a specific directory.

**Comment by samgdotson** [comment]
@LukeSeifert I think I have addressed your questions! Thanks for the review.

**Review by LukeSeifert — APPROVED** [review]
Thanks for addressing my comments and questions! Looks good, let me know if you want me to merge or if you'd rather wait for more reviews.

**Comment by samgdotson** [comment]
@LukeSeifert you can merge when ready

**Pull request merged**